### DISCUSSION TITLE
Heightfieldshape and streaming terrain

### DISCUSSION THREAD

**Discussion by Srekel** [jrouwe/JoltPhysics#523] [Q&A] [answered]
Hi,

I was wondering what your suggestion on this is. In our game Tides of Revival a patch is 2^n big, depending on the patch. Right now the resolution for every patch is 65 units, and the lowest (best) lod, lod0, is 64m wide. So the number of samples is 65 (the patch images containing the heightmaps are 65x65, even for higher (worse) lods).

Now, I'm not interested in lodding physics, I think that is never going to work well :) Physics will always be lod0.

But, I do feel like the physics ought to match the visuals at lod0, and the rendered terrain at lod0 is then at a 1m resolution (64/64). But using the HeightFieldShape it will need to be 64 samples, thus the resolution will be slightly less (1.02 sample/m). This should mean that just before halfway the physics sample will potentially be off from the rendered sample by about 0.5 meters.

Worst case that this could result in is: You're looking for the murder weapon to solve the crime but can't see it because it happens to be just under a fold in the terrain.

I see a few solutions...

1. Live with it.
2. Use double physics resolution. Still live with it but presumably less likely to cause issues. Potentially do this dynamically per-patch based on the roughness of the terrain.
3. Use a mesh shape
4. Use a heightfield but let it span 63 meters and use a mesh shape to "stitch" the missing meter.
5. Switch to patches that are 2^n-1 wide. (I feel this would complicate a bunch of other things so would prefer not to..)
6. ... maybe something else I'm missing? I feel like this is a common use case for streamed terrain but not sure how it's typically solved.

This is a video showing the game: https://youtu.be/t7dfbwfrX7Q
Related question: https://github.com/jrouwe/JoltPhysics/discussions/400
Tides of Revival repo: https://github.com/Srekel/elvengroin-legacy

**Comment by Srekel** [comment]
It would be interesting if it would be possible to ballpark the efficiency - would a heightfield at 2x resolution potentially be faster than a mesh shape?

**Comment by jrouwe** [comment, marked as answer]
Hello,

I don't think there is a massive performance difference between a MeshShape and a HeightFieldShape (haven't measured it in this context though, just a gut feel). The main reason why the HeightFieldShape exists is because it uses much less memory, but those memory savings only add up when you're using bigger height fields (e.g. 256x256 or more). For 65x65 the savings will be less, so I'm not sure how much benefit you're getting from using height fields.

I think the options you list are the options there are. There is 1 additional option which involves setting HeightFieldShapeSettings::mBlockSize to an odd value to get closer to 65 samples. E.g. mBlockSize = 5 and number of samples = 5*16 = 80 (wastes 80-65 = 15 samples on each side which need to be set to cNoCollisionValue).

**Reply by Srekel** [reply]
Ok, thanks! If so I might just go with a mesh shape for now, but I'll keep your suggestion in mind for the future. I didn't realize parts of the shape could be marked as uncollidable :)

**Comment by Srekel** [comment]
Oh @jrouwe I have a follow-up question :) I was just wondering if you'd be able to tell me what Horizon uses for it's streaming terrain. Just in basic terms of tile size and how you conceptually set up the shapes for the tiles. But totally understand if it's top secret stuff!

**Reply by jrouwe** [reply]
There is a little bit of info [here](http://jrouwe.nl/architectingjolt/ArchitectingJoltPhysics_Rouwe_Jorrit_Notes.pdf). I believe we use a sample every 0.5m.

**Reply by Srekel** [reply]
Ah yes I found that earlier while researching this question :) But it didn't explain if you use the heightfieldshape and if so, if you did anything in particular to make it align with the rendered ground or if you just stretched it and that was good enough. (and honestly if it's good enough for Horizon it *probably* is for us too! :D )

**Reply by jrouwe** [reply]
Yes we use the HeightFieldShape for terrain and the rendered ground uses the exact same resolution and dimensions as physics. I think there are 512x512 samples per tile so there are 511x511 quads which are then stretched to 256x256m. The height fields are made so that the border row/column matches the neighboring tile so that there are no seams.
